File: gcj2014R1BA.hs
#!/usr/bin/env runghc
import Data.List
import Data.Function
main = getContents >>= mapM_ putStrLn . zipWith caseno [1..] . solve . lines
caseno i r = "Case #"++show i++": "++r
solve (c:cs) = map (toString . answer) . take (read c) . analyze $ cs ::[String]
analyze (c:cs) = ss:analyze r
  where (ss,r) = splitAt (read c) cs
toString = id
answer ss = if c then "Fegla Won" else show $ sum' (map f wss) where
  wss = transpose . map (map (\w->(head w,length w)).group) $ ss
  c = c1 || c2 where
    c1 = ((/=) `on` length) (head wss) (last wss)
    c2 = any (\ws->any (((/=) `on` fst) $ head ws) ws) wss
f ws = sum' $ map (abs . subtract b) ns where
  ns = sort $ map snd ws
  b = ns!!(length ns`div`2)
sum' = foldl' (+) 0

File: gcj2014R1BA.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <limits>

static auto solve = []()
{
    int N; std::cin >> N;
    std::vector<std::string> S(N);
    for(auto& v: S) std::cin >> v;
    auto cs = S[0];
    cs.erase(std::unique(cs.begin(),cs.end()),cs.end());
    int n=0;
    std::vector<int> p(N,0);
    for(auto c: cs)
    {
        std::vector<int> cnt(N,0);
        for(int i=0; i<N; ++i)
        {
            const auto& s = S[i];
            const int sn = s.size();
            int j=p[i];
            for( ; j<sn && s[j]==c; ++j);
            if((cnt[i]=j-p[i])==0)
                return std::string("Fegla Won");
            p[i]=j;
        }
        std::sort(cnt.begin(),cnt.end());
        for(int i=0; i<N; ++i)
            n+=std::abs(cnt[i]-cnt[N/2]);
    }
    for(int i=0; i<N; ++i)
    {
        const int sn = S[i].size();
        if(p[i]!=sn)
            return std::string("Fegla Won");
    }
    return std::to_string(n);
};

int main()
{
    std::cin.tie(0);
    std::ios_base::sync_with_stdio(false);
    int T; std::cin >> T;
    for (int i=1; i<=T; ++i)
        std::cout << "Case #" << i << ": " << solve() << std::endl;
}
